File: the-versatile-sieve-of-eratosthenes/EratosthenesSievesTest.cpp
/* 
 * Test of variants of the sieve of Eratosthenes (C++)
 * by Project Nayuki, 2017. Public domain.
 * https://www.nayuki.io/page/the-versatile-sieve-of-eratosthenes
 */

#include <cstddef>
#include <cstdint>
#include <cstdlib>
#include <iostream>
#include <vector>
#include "EratosthenesSieves.hpp"

using std::size_t;
using std::uint32_t;
using std::vector;


// Forward declarations
static void testValues();
static void testPrefixConsistency();


int main() {
	try {
		testValues();
		testPrefixConsistency();
		
		std::cerr << "Test passed" << std::endl;
		return EXIT_SUCCESS;
	} catch (const char *msg) {
		std::cerr << msg << std::endl;
		return EXIT_FAILURE;
	}
}


static void testValues() {
	{
		vector<bool> expected{false, false, true, true, false, true, false, true, false, false, false, true, false, true, false, false, false, true, false, true, false, false, false, true, false, false, false, false, false, true, false};
		vector<bool> actual = sievePrimeness(30);
		if (actual != expected)
			throw "Mismatch";
	}
	{
		vector<uint32_t> expected{0, 1, 2, 3, 2, 5, 2, 7, 2, 3, 2, 11, 2, 13, 2, 3, 2, 17, 2, 19, 2, 3, 2, 23, 2, 5, 2, 3, 2, 29, 2};
		vector<uint32_t> actual = sieveSmallestPrimeFactor(30);
		if (actual != expected)
			throw "Mismatch";
	}
	{
		vector<uint32_t> expected{0, 1, 1, 2, 2, 4, 2, 6, 4, 6, 4, 10, 4, 12, 6, 8, 8, 16, 6, 18, 8, 12, 10, 22, 8, 20, 12, 18, 12, 28, 8};
		vector<uint32_t> actual = sieveTotient(30);
		if (actual != expected)
			throw "Mismatch";
	}
	{
		vector<uint32_t> expected{0, 0, 1, 1, 1, 1, 2, 1, 1, 1, 2, 1, 2, 1, 2, 2, 1, 1, 2, 1, 2, 2, 2, 1, 2, 1, 2, 1, 2, 1, 3};
		vector<uint32_t> actual = sieveOmega(30);
		if (actual != expected)
			throw "Mismatch";
	}
	{
		vector<uint32_t> expected{0, 1, 2, 3, 2, 5, 6, 7, 2, 3, 10, 11, 6, 13, 14, 15, 2, 17, 6, 19, 10, 21, 22, 23, 6, 5, 26, 3, 14, 29, 30};
		vector<uint32_t> actual = sieveRadical(30);
		if (actual != expected)
			throw "Mismatch";
	}
}


static void testPrefixConsistency() {
	const uint32_t N = 10000;
	{
		vector<bool> prev;
		for (uint32_t i = 0; i < N; i++) {
			vector<bool> cur = sievePrimeness(i);
			for (uint32_t j = 0; j < i; j++) {
				if (cur.at(j) != prev.at(j))
					throw "Mismatch";
			}
			prev = cur;
		}
	}
	{
		vector<vector<uint32_t> (*)(uint32_t)> FUNCS{
			sieveSmallestPrimeFactor,
			sieveTotient,
			sieveOmega,
			sieveRadical,
		};
		for (vector<uint32_t> (*func)(uint32_t) : FUNCS) {
			vector<uint32_t> prev;
			for (uint32_t i = 0; i < N; i++) {
				vector<uint32_t> cur = func(i);
				for (uint32_t j = 0; j < i; j++) {
					if (cur.at(j) != prev.at(j))
						throw "Mismatch";
				}
				prev = cur;
			}
		}
	}
}

File: the-versatile-sieve-of-eratosthenes/eratosthenessieves-test.rs
/* 
 * Test of variants of the sieve of Eratosthenes (Rust)
 * by Project Nayuki, 2017. Public domain.
 * https://www.nayuki.io/page/the-versatile-sieve-of-eratosthenes
 */

mod eratosthenessieves;


fn main() {
	test_values();
	test_prefix_consistency(eratosthenessieves::sieve_primeness);
	test_prefix_consistency(eratosthenessieves::sieve_smallest_prime_factor);
	test_prefix_consistency(eratosthenessieves::sieve_totient);
	test_prefix_consistency(eratosthenessieves::sieve_omega);
	test_prefix_consistency(eratosthenessieves::sieve_radical);
	println!("Test passed");
}


fn test_values() {
	let expect = vec![false, false, true, true, false, true, false, true, false, false, false, true, false, true, false, false, false, true, false, true, false, false, false, true, false, false, false, false, false, true, false];
	let actual = eratosthenessieves::sieve_primeness((expect.len() - 1) as u32);
	assert_equals(&expect, &actual);
	
	let expect = vec![0u32, 1, 2, 3, 2, 5, 2, 7, 2, 3, 2, 11, 2, 13, 2, 3, 2, 17, 2, 19, 2, 3, 2, 23, 2, 5, 2, 3, 2, 29, 2];
	let actual = eratosthenessieves::sieve_smallest_prime_factor((expect.len() - 1) as u32);
	assert_equals(&expect, &actual);
	
	let expect = vec![0u32, 1, 1, 2, 2, 4, 2, 6, 4, 6, 4, 10, 4, 12, 6, 8, 8, 16, 6, 18, 8, 12, 10, 22, 8, 20, 12, 18, 12, 28, 8];
	let actual = eratosthenessieves::sieve_totient((expect.len() - 1) as u32);
	assert_equals(&expect, &actual);
	
	let expect = vec![0u32, 0, 1, 1, 1, 1, 2, 1, 1, 1, 2, 1, 2, 1, 2, 2, 1, 1, 2, 1, 2, 2, 2, 1, 2, 1, 2, 1, 2, 1, 3];
	let actual = eratosthenessieves::sieve_omega((expect.len() - 1) as u32);
	assert_equals(&expect, &actual);
	
	let expect = vec![0u32, 1, 2, 3, 2, 5, 6, 7, 2, 3, 10, 11, 6, 13, 14, 15, 2, 17, 6, 19, 10, 21, 22, 23, 6, 5, 26, 3, 14, 29, 30];
	let actual = eratosthenessieves::sieve_radical((expect.len() - 1) as u32);
	assert_equals(&expect, &actual);
}


fn test_prefix_consistency<T: std::cmp::PartialEq>(func: fn(u32) -> Vec<T>) {
	let n: u32 = 10000;
	let mut prev = Vec::<T>::new();
	for i in 0 .. n {
		let cur: Vec<T> = func(i);
		for j in 0 .. i {
			assert_equals(&prev[j as usize], &cur[j as usize]);
		}
		prev = cur;
	}
}


fn assert_equals<T: std::cmp::PartialEq>(expect: &T, actual: &T) {
	if actual != expect {
		panic!("Mismatch");
	}
}
